### PR TITLE
[Clang] Improve testing for the flexible array member

### PR DESCRIPTION
**Pull request by bwendling**
Testing for the name of the flexible array member isn't as robust as testing the FieldDecl pointers.

### CODE REVIEW

**Comment by kees** [comment]
Does this still work for cases where there are multiple flexible arrays? e.g.

```
struct weird_protocol {
    unsigned int cmd_type;
    unsigned int data_len;
    union {
        struct cmd_one one[];
        struct cmd_two two[];
        struct cmd_three three[];
        unsigned char raw_bytes[];
    };
};
```

**Comment by bwendling** [comment]
> Does this still work for cases where there are multiple flexible arrays? e.g.
> 
> ```
> struct weird_protocol {
>     unsigned int cmd_type;
>     unsigned int data_len;
>     union {
>         struct cmd_one one[];
>         struct cmd_two two[];
>         struct cmd_three three[];
>         unsigned char raw_bytes[];
>     };
> };
> ```

Ah! One of your heresies. :-)

The current code probably doesn't support this (I'll check). This PR shouldn't change this. I'll have to revisit this code to make sure that it can be supported.

**Comment by efriedma-quic** [comment]
Strictly speaking, I don't think this is NFC.  Consider, for example:

```
struct S {
  struct X { int array[1]; } x;
  struct Y { int count; int array[] __attribute__((__counted_by__(count))); } y;
};
void f(void* __attribute__((pass_dynamic_object_size(0))));
void g(struct S* s) { f(s->y.array); }
```

**Comment by bwendling** [comment]
> Strictly speaking, I don't think this is NFC. Consider, for example:
> 
> ```
> struct S {
>   struct X { int array[1]; } x;
>   struct Y { int count; int array[] __attribute__((__counted_by__(count))); } y;
> };
> void f(void* __attribute__((pass_dynamic_object_size(0))));
> void g(struct S* s) { f(s->y.array); }
> ```

Hmm.. Yeah, this change makes it more accurate. I'll remove the tag.

**Comment by bwendling** [comment]
Would changing the `decl_iterator` into a bidirectional iterator be acceptable? That way I could grab the last `FieldDecl` to check if it's a FAM rather than iterating through unnecessary `Decls`...

**Comment by efriedma-quic** [comment]
If you can pull it off without increasing the size of Decl, sure; not sure it's worthwhile otherwise.

File: clang/lib/CodeGen/CGBuiltin.cpp [hunk 1]
```diff
@@ -930,12 +931,12 @@ CodeGenFunction::emitFlexibleArrayMemberSize(const Expr *E, unsigned Type,
...
   // Get the flexible array member Decl.
   const RecordDecl *OuterRD = nullptr;
-  std::string FAMName;
+  const FieldDecl *FAMDecl = nullptr;
   if (const auto *ME = dyn_cast<MemberExpr>(Base)) {
     // Check if \p Base is referencing the FAM itself.
     const ValueDecl *VD = ME->getMemberDecl();
     OuterRD = VD->getDeclContext()->getOuterLexicalRecordContext();
-    FAMName = VD->getNameAsString();
+    FAMDecl = dyn_cast<FieldDecl>(VD);
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/CGBuiltin.cpp:939]: If VD isn't a FieldDecl, should we return early here?

**Comment by bwendling** [comment]
A quick look at `Decl` seems like it would require adding a `PrevInContextAndBits`, which would increase the size by `sizeof(Decl *)`. I might be able to massage a way to make it work, but it's probably not worth it.

File: clang/lib/CodeGen/CGBuiltin.cpp (hunk 1 above)
Comment (bwendling) [reply clang/lib/CodeGen/CGBuiltin.cpp:939]: Wouldn't hurt. Done.

**Review by efriedma-quic — APPROVED** [review]
LGTM

**Pull request merged**